File: scripts/migrate-all.ts
// Full migration: local workspace + GitHub repo -> RAG + mem0
// RAG gets chunks (6K), mem0 gets whole files (LLM extraction handles it)
// bun run scripts/migrate-all.ts

import { readdir, readFile } from "fs/promises"
import { join, basename, extname } from "path"
import { Database } from "bun:sqlite"

const RAG = "http://localhost:3847"
const MEM0 = "http://localhost:3848"
const TAG = "openclaw_Joaos_MacBook_Pro_local"
const HOME = process.env.HOME!
const LOCAL_MEM = join(HOME, ".openclaw/workspace/memory")
const LOCAL_WS = join(HOME, ".openclaw/workspace")
const GITHUB_MEM = "/tmp/my-clawdbot/workspace/memory"
const GITHUB_WS = "/tmp/my-clawdbot/workspace"
const CHROMA = join(HOME, ".openclaw/mem0/data/chroma/chroma.sqlite3")
const CHUNK_SIZE = 6000
const MEM0_MAX_TEXT = 50000  // mem0 can handle larger texts via LLM
const RAG_WORKERS = 5
const MEM0_WORKERS = 10
const MAX_RETRIES = 3

type Item = { sid: string; text: string; date?: string; label?: string }
type Chunk = { sid: string; text: string; date?: string; label?: string }

const stats = { ragOk: 0, ragFail: 0, mem0Ok: 0, mem0Fail: 0 }

async function retry<T>(fn: () => Promise<T>): Promise<T> {
  for (let i = 0; i < MAX_RETRIES; i++) {
    try { return await fn() }
    catch (e) {
      if (i === MAX_RETRIES - 1) throw e
      await new Promise(r => setTimeout(r, 2000 * (i + 1)))
    }
  }
  throw new Error("unreachable")
}

async function post(url: string, body: unknown): Promise<void> {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), 600_000)
  try {
    const res = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
      signal: controller.signal,
    })
    if (!res.ok) {
      const t = await res.text().catch(() => "")
      throw new Error(`${res.status}: ${t.slice(0, 200)}`)
    }
  } finally {
    clearTimeout(timer)
  }
}

function chunkForRag(text: string): string[] {
  if (text.length <= CHUNK_SIZE) return [text]
  const out: string[] = []
  let i = 0
  while (i < text.length) {
    let end = Math.min(i + CHUNK_SIZE, text.length)
    if (end < text.length) {
      const nl = text.lastIndexOf("\n", end)
      if (nl > i + CHUNK_SIZE * 0.4) end = nl + 1
    }
    out.push(text.slice(i, end))
    i = end
  }
  return out
}

// for mem0, just truncate if too long
function trimForMem0(text: string): string {
  return text.length <= MEM0_MAX_TEXT ? text : text.slice(0, MEM0_MAX_TEXT)
}

function getDate(f: string): string | undefined {
  return f.match(/^(\d{4}-\d{2}-\d{2})/)?.[1]
}

async function loadDir(dir: string, exts: string[]): Promise<{ path: string; name: string }[]> {
  try {
    const entries = await readdir(dir)
    return entries
      .filter(f => exts.includes(extname(f).toLowerCase()))
      .sort()
      .map(f => ({ path: join(dir, f), name: basename(f, extname(f)) }))
  } catch { return [] }
}

async function loadChromaFacts(): Promise<string[]> {
  try {
    const db = new Database(CHROMA, { readonly: true })
    const rows = db.query<{ data: string }, []>(`
      SELECT em_data.string_value as data
      FROM embedding_metadata em_uid
      JOIN embedding_metadata em_data ON em_uid.id = em_data.id AND em_data.key = 'data'
      JOIN embedding_metadata em_src ON em_uid.id = em_src.id AND em_src.key = 'source'
      WHERE em_uid.key = 'user_id' AND em_uid.string_value = 'jinstronda'
      AND em_src.string_value = 'openclaw'
    `).all()
    db.close()
    return rows.map(r => r.data).filter(Boolean)
  } catch { return [] }
}

// RAG worker: processes chunks from shared index
async function ragWorker(chunks: Chunk[], idx: { v: number }) {
  while (true) {
    const i = idx.v++
    if (i >= chunks.length) break
    const c = chunks[i]
    try {
      await retry(() => post(`${RAG}/ingest`, {
        containerTag: TAG,
        sessionId: c.sid,
        messages: [{ role: "user", content: c.text }],
        ...(c.date ? { date: c.date } : {}),
      }))
      stats.ragOk++
    } catch (e) {
      stats.ragFail++
      console.log(`  FAIL rag [${c.label || c.sid}] ${e}`)
    }
  }
}

// mem0 worker: processes whole items from shared index
async function mem0Worker(items: Item[], idx: { v: number }) {
  while (true) {
    const i = idx.v++
    if (i >= items.length) break
    const item = items[i]
    try {
      const text = trimForMem0(item.text)
      await retry(() => post(`${MEM0}/add`, { text, user_id: TAG }))
      stats.mem0Ok++
    } catch (e) {
      stats.mem0Fail++
      console.log(`  FAIL mem0 [${item.label || item.sid}] ${e}`)
    }
  }
}

const MEM0_ONLY = process.argv.includes("--mem0-only")
const RAG_ONLY = process.argv.includes("--rag-only")

async function main() {
  if (!MEM0_ONLY) {
    const rh = await fetch(`${RAG}/health`).catch(() => null)
    if (!rh?.ok) { console.log("RAG not ready"); process.exit(1) }
  }
  if (!RAG_ONLY) {
    const mh = await fetch(`${MEM0}/health`).catch(() => null)
    if (!mh?.ok) { console.log("mem0 not ready"); process.exit(1) }
  }

  console.log(`RAG: ${RAG_WORKERS} workers, ${CHUNK_SIZE} chunk | mem0: ${MEM0_WORKERS} workers, whole files`)
  console.log(`RAG: ${RAG} | mem0: ${MEM0} | tag: ${TAG}\n`)

  const seen = new Set<string>()
  const h = (t: string) => Bun.hash(t.trim().toLowerCase()).toString(36)
  const items: Item[] = []
  let skipCount = 0

  // 1. Identity files
  const idFiles = ["MEMORY.md", "IDENTITY.md", "SOUL.md", "USER.md", "AGENTS.md", "HEARTBEAT.md", "HORMOZI.md", "TOOLS.md"]
  let idCount = 0
  for (const dir of [LOCAL_WS, GITHUB_WS]) {
    for (const f of idFiles) {
      try {
        const content = await readFile(join(dir, f), "utf-8")
        const key = h(content)
        if (seen.has(key)) continue
        seen.add(key)
        idCount++
        items.push({ sid: `id_${basename(f, ".md")}`, text: content, date: "2026-02-23", label: f })
      } catch {}
    }
  }
  console.log(`[1/4] ${idCount} identity files`)

  // 2. Local memory files
  const localFiles = await loadDir(LOCAL_MEM, [".md", ".txt"])
  let localCount = 0
  for (const f of localFiles) {
    try {
      const content = await readFile(f.path, "utf-8")
      const key = h(content)
      if (seen.has(key)) { skipCount++; continue }
      seen.add(key)
      localCount++
      items.push({ sid: `local_${f.name}`, text: content, date: getDate(basename(f.path)), label: f.name })
    } catch {}
  }
  console.log(`[2/4] ${localCount} local memory files`)

  // 3. GitHub repo memory files
  const ghFiles = await loadDir(GITHUB_MEM, [".md", ".txt"])
  let ghCount = 0
  for (const f of ghFiles) {
    try {
      const content = await readFile(f.path, "utf-8")
      const key = h(content)
      if (seen.has(key)) { skipCount++; continue }
      seen.add(key)
      ghCount++
      items.push({ sid: `gh_${f.name}`, text: content, date: getDate(basename(f.path)), label: `gh:${f.name}` })
    } catch {}
  }
  console.log(`[3/4] ${ghCount} GitHub memory files (${skipCount} dupes skipped)`)

  // 4. Chroma openclaw facts
  const facts = await loadChromaFacts()
  let factCount = 0
  for (let i = 0; i < facts.length; i++) {
    const key = h(facts[i])
    if (seen.has(key)) continue
    seen.add(key)
    factCount++
    items.push({ sid: `chroma_${i}`, text: facts[i], label: `chroma:${i}` })
  }
  console.log(`[4/4] ${factCount} Chroma facts`)

  // chunk items for RAG only
  const ragChunks: Chunk[] = []
  for (const item of items) {
    if (!item.text.trim()) continue
    const chunks = chunkForRag(item.text)
    for (let c = 0; c < chunks.length; c++) {
      ragChunks.push({
        sid: chunks.length === 1 ? item.sid : `${item.sid}_c${c}`,
        text: chunks[c],
        date: item.date,
        label: item.label,
      })
    }
  }

  const totalItems = idCount + localCount + ghCount + factCount
  console.log(`\nTotal: ${totalItems} items`)
  console.log(`  RAG: ${ragChunks.length} chunks (${RAG_WORKERS} workers)`)
  console.log(`  mem0: ${items.length} whole files (${MEM0_WORKERS} workers, serialized on server)\n`)

  const t0 = performance.now()

  const interval = setInterval(() => {
    const elapsed = ((performance.now() - t0) / 1000).toFixed(0)
    const ragPct = ((stats.ragOk / ragChunks.length) * 100).toFixed(1)
    const mem0Pct = ((stats.mem0Ok / items.length) * 100).toFixed(1)
    console.log(`  [${elapsed}s] rag: ${stats.ragOk}/${ragChunks.length} (${ragPct}%) ${stats.ragFail}f | mem0: ${stats.mem0Ok}/${items.length} (${mem0Pct}%) ${stats.mem0Fail}f`)
  }, 10_000)

  const ragIdx = { v: 0 }
  const mem0Idx = { v: 0 }

  const tasks: Promise<void[]>[] = []
  if (!MEM0_ONLY) tasks.push(Promise.all(Array.from({ length: RAG_WORKERS }, () => ragWorker(ragChunks, ragIdx))))
  if (!RAG_ONLY) tasks.push(Promise.all(Array.from({ length: MEM0_WORKERS }, () => mem0Worker(items, mem0Idx))))
  await Promise.all(tasks)

  clearInterval(interval)
  const elapsed = ((performance.now() - t0) / 1000).toFixed(1)
  console.log(`\nDone in ${elapsed}s`)
  console.log(`  RAG:  ${stats.ragOk} ok, ${stats.ragFail} fail (${ragChunks.length} chunks)`)
  console.log(`  mem0: ${stats.mem0Ok} ok, ${stats.mem0Fail} fail (${items.length} items)`)
  console.log(`  Deduped: ${skipCount}`)
}

main()

File: scripts/migrate-all.py
#!/usr/bin/env python3
"""Migrate ALL OpenClaw memories into dual-layer system (RAG + mem0 + graph).

RAG gets chunked text (6K). mem0 gets whole files (LLM extraction handles it).
Separate concurrency for each: RAG is fast, mem0 is slow (LLM per call).

Sources:
  1. Identity files from local + GitHub workspace
  2. Memory .md/.txt from local + GitHub workspace/memory
  3. Chroma DB openclaw-source facts

Usage:
    PYTHONUNBUFFERED=1 python3 scripts/migrate-all.py
"""
from __future__ import annotations

import asyncio
import hashlib
import os
import re
import sqlite3
import sys
import time
from pathlib import Path

import aiohttp

RAG_URL = os.environ.get("RAG_URL", "http://localhost:3847")
MEM0_URL = os.environ.get("MEM0_URL", "http://localhost:3848")
TAG = "openclaw_Joaos_MacBook_Pro_local"

LOCAL_WS = Path.home() / ".openclaw/workspace"
LOCAL_MEM = LOCAL_WS / "memory"
GITHUB_WS = Path("/tmp/my-clawdbot/workspace")
GITHUB_MEM = GITHUB_WS / "memory"
CHROMA_DB = Path.home() / ".openclaw/mem0/data/chroma/chroma.sqlite3"

MAX_CHUNK = 6000
MEM0_MAX_TEXT = 50000
RAG_CONCURRENCY = 30
MEM0_CONCURRENCY = 5

IDENTITY_FILES = ["MEMORY.md", "IDENTITY.md", "SOUL.md", "USER.md", "AGENTS.md", "HEARTBEAT.md", "HORMOZI.md", "TOOLS.md"]

stats = {"rag_ok": 0, "rag_fail": 0, "mem0_ok": 0, "mem0_fail": 0}


def chunk_text(text: str) -> list[str]:
    if len(text) <= MAX_CHUNK:
        return [text]
    out = []
    i = 0
    while i < len(text):
        end = min(i + MAX_CHUNK, len(text))
        if end < len(text):
            nl = text.rfind("\n", i + int(MAX_CHUNK * 0.4), end)
            if nl > 0:
                end = nl + 1
        out.append(text[i:end])
        i = end
    return out


def extract_date(filename: str) -> str | None:
    m = re.match(r"(\d{4}-\d{2}-\d{2})", filename)
    return m.group(1) if m else None


def content_hash(text: str) -> str:
    return hashlib.md5(text.strip().lower().encode()).hexdigest()


def load_chroma_facts() -> list[str]:
    if not CHROMA_DB.exists():
        return []
    try:
        conn = sqlite3.connect(str(CHROMA_DB))
        rows = conn.execute("""
            SELECT em_data.string_value
            FROM embedding_metadata em_uid
            JOIN embedding_metadata em_data ON em_uid.id = em_data.id AND em_data.key = 'data'
            JOIN embedding_metadata em_src ON em_uid.id = em_src.id AND em_src.key = 'source'
            WHERE em_uid.key = 'user_id' AND em_uid.string_value = 'jinstronda'
            AND em_src.string_value = 'openclaw'
        """).fetchall()
        conn.close()
        return [r[0] for r in rows if r[0]]
    except Exception as e:
        print(f"Chroma read failed: {e}")
        return []


async def rag_worker(session: aiohttp.ClientSession, queue: asyncio.Queue):
    timeout = aiohttp.ClientTimeout(total=120)
    while True:
        item = await queue.get()
        if item is None:
            queue.task_done()
            break
        sid, text, date = item
        body = {
            "containerTag": TAG,
            "sessionId": sid,
            "messages": [{"role": "user", "content": text}],
        }
        if date:
            body["date"] = date
        for attempt in range(3):
            try:
                async with session.post(f"{RAG_URL}/ingest", json=body, timeout=timeout) as resp:
                    if resp.status == 200:
                        stats["rag_ok"] += 1
                        break
                    if attempt == 2:
                        stats["rag_fail"] += 1
            except Exception:
                if attempt == 2:
                    stats["rag_fail"] += 1
                else:
                    await asyncio.sleep(1 * (attempt + 1))
        queue.task_done()


async def mem0_worker(session: aiohttp.ClientSession, queue: asyncio.Queue):
    timeout = aiohttp.ClientTimeout(total=180)
    while True:
        item = await queue.get()
        if item is None:
            queue.task_done()
            break
        sid, text = item
        text = text[:MEM0_MAX_TEXT]
        body = {"text": text, "user_id": TAG}
        for attempt in range(3):
            try:
                async with session.post(f"{MEM0_URL}/add", json=body, timeout=timeout) as resp:
                    if resp.status == 200:
                        stats["mem0_ok"] += 1
                        break
                    if attempt == 2:
                        stats["mem0_fail"] += 1
                        print(f"  FAIL mem0 [{sid}] status={resp.status}")
            except Exception as e:
                if attempt == 2:
                    stats["mem0_fail"] += 1
                    print(f"  FAIL mem0 [{sid}] {e}")
                else:
                    await asyncio.sleep(2 * (attempt + 1))
        queue.task_done()


async def main():
    async with aiohttp.ClientSession() as session:
        for url, name in [(RAG_URL, "RAG"), (MEM0_URL, "mem0")]:
            try:
                async with session.get(f"{url}/health") as r:
                    assert r.status == 200
            except Exception:
                print(f"{name} not reachable at {url}")
                sys.exit(1)

        print(f"RAG: {RAG_URL} ({RAG_CONCURRENCY}w) | mem0: {MEM0_URL} ({MEM0_CONCURRENCY}w)")
        print(f"tag: {TAG}\n")

        seen = set()
        items: list[tuple[str, str, str | None]] = []  # (sid, text, date)
        skip_count = 0

        # 1. identity files
        id_count = 0
        for ws_dir in [LOCAL_WS, GITHUB_WS]:
            for fname in IDENTITY_FILES:
                fp = ws_dir / fname
                if not fp.exists():
                    continue
                content = fp.read_text(errors="replace")
                h = content_hash(content)
                if h in seen:
                    skip_count += 1
                    continue
                seen.add(h)
                id_count += 1
                items.append((f"id_{fp.stem}", content, "2026-02-23"))
        print(f"[1/4] {id_count} identity files")

        # 2. local memory files
        local_count = 0
        if LOCAL_MEM.exists():
            for fp in sorted(LOCAL_MEM.iterdir()):
                if fp.suffix.lower() not in (".md", ".txt") or not fp.is_file():
                    continue
                content = fp.read_text(errors="replace")
                h = content_hash(content)
                if h in seen:
                    skip_count += 1
                    continue
                seen.add(h)
                local_count += 1
                items.append((f"local_{fp.stem}", content, extract_date(fp.name)))
        print(f"[2/4] {local_count} local memory files")

        # 3. github memory files
        gh_count = 0
        if GITHUB_MEM.exists():
            for fp in sorted(GITHUB_MEM.iterdir()):
                if fp.suffix.lower() not in (".md", ".txt") or not fp.is_file():
                    continue
                content = fp.read_text(errors="replace")
                h = content_hash(content)
                if h in seen:
                    skip_count += 1
                    continue
                seen.add(h)
                gh_count += 1
                items.append((f"gh_{fp.stem}", content, extract_date(fp.name)))
        print(f"[3/4] {gh_count} GitHub memory files ({skip_count} dupes skipped)")

        # 4. chroma facts
        facts = load_chroma_facts()
        fact_count = 0
        for j, fact in enumerate(facts):
            h = content_hash(fact)
            if h in seen:
                continue
            seen.add(h)
            fact_count += 1
            items.append((f"chroma_{j}", fact, None))
        print(f"[4/4] {fact_count} Chroma facts")

        # build RAG chunk queue and mem0 file queue
        rag_queue: asyncio.Queue = asyncio.Queue()
        mem0_queue: asyncio.Queue = asyncio.Queue()

        rag_chunk_count = 0
        for sid, text, date in items:
            if not text.strip():
                continue
            chunks = chunk_text(text)
            for ci, chunk in enumerate(chunks):
                csid = sid if len(chunks) == 1 else f"{sid}_c{ci}"
                rag_queue.put_nowait((csid, chunk, date))
                rag_chunk_count += 1
            mem0_queue.put_nowait((sid, text))

        total_items = id_count + local_count + gh_count + fact_count
        mem0_count = mem0_queue.qsize()
        print(f"\nTotal: {total_items} items")
        print(f"  RAG: {rag_chunk_count} chunks ({RAG_CONCURRENCY} workers)")
        print(f"  mem0: {mem0_count} whole files ({MEM0_CONCURRENCY} workers)")
        print()

        # sentinel values to stop workers
        for _ in range(RAG_CONCURRENCY):
            rag_queue.put_nowait(None)
        for _ in range(MEM0_CONCURRENCY):
            mem0_queue.put_nowait(None)

        t0 = time.time()

        # progress reporter
        async def progress():
            while True:
                await asyncio.sleep(10)
                elapsed = int(time.time() - t0)
                rag_pct = (stats["rag_ok"] / max(rag_chunk_count, 1)) * 100
                m0_pct = (stats["mem0_ok"] / max(mem0_count, 1)) * 100
                print(
                    f"  [{elapsed}s] rag: {stats['rag_ok']}/{rag_chunk_count} ({rag_pct:.1f}%) {stats['rag_fail']}f"
                    f" | mem0: {stats['mem0_ok']}/{mem0_count} ({m0_pct:.1f}%) {stats['mem0_fail']}f"
                )

        progress_task = asyncio.create_task(progress())

        # run workers
        rag_workers = [asyncio.create_task(rag_worker(session, rag_queue)) for _ in range(RAG_CONCURRENCY)]
        mem0_workers = [asyncio.create_task(mem0_worker(session, mem0_queue)) for _ in range(MEM0_CONCURRENCY)]

        await asyncio.gather(*rag_workers, *mem0_workers)
        progress_task.cancel()

        elapsed = time.time() - t0
        print(f"\nDone in {elapsed:.1f}s")
        print(f"  RAG:  {stats['rag_ok']} ok, {stats['rag_fail']} fail ({rag_chunk_count} chunks)")
        print(f"  mem0: {stats['mem0_ok']} ok, {stats['mem0_fail']} fail ({mem0_count} items)")
        print(f"  Deduped: {skip_count}")


if __name__ == "__main__":
    asyncio.run(main())
